### DISCUSSION TITLE
any interest in geopandas/matplotlib plotting option?

### DISCUSSION THREAD

**Discussion by everettsp** [pyswmm/swmmio#227] [Ideas]
Would this type of plotting function be a useful addition to swmmio? If so, I can polish it up and submit a PR. Would also make it easy to export to GeoJSON, which I see hasn't been implemented yet.

![image](https://github.com/user-attachments/assets/715fcd9b-8b7d-47c1-b080-11a1ccffc284)

**Comment by bemcdonnell** [comment]
@everettsp, I think it would fit with SWMMIO the best. Have a look at the viz library in there. My suggestion is to make it fairly modular since we might move all viz out of that tool later

**Comment by aerispaha** [comment]
@everettsp, I like this idea too. Keep in mind though that there already is similar functionality provided in the `swmmio.graphics` module. With that you can get an visual representation of the model in a jupyter notebook, exported as PNG, or as an interactive map in an HTML file. 

```python
import swmmio

# create a swmmio Model object with a coordinate reference system 
model = swmmio.Model('path.inp', crs='EPSG:4326')

# create an image of the model 
swmmio.draw_model(model)

# visualize the model in an interactive map 
swmmio.create_map(model, filename="my-cool-model-map.html")
```
![image](https://github.com/user-attachments/assets/03c9bd5b-7a34-41c7-8f9d-63b3aaaf8131)


That said, the `draw_model` function uses Pillow for graphics, and I think it would be a lot more sensible to use matplotlib, or matplotlib via Geopandas. I also really would love to be able to have the interactive maps be compatible with Jupyter notebooks, similar to [folium](https://python-visualization.github.io/folium/latest/getting_started.html#Creating-a-map) and/or how the Geopandas [explore()](https://geopandas.org/en/stable/docs/user_guide/interactive_mapping.html#Interactive-mapping) method works. This is something that's been vexing me for a while.

We can also access the wealth of Geopandas / matplotlib plotting utilities by tapping into the `swmmio.elements` API, which aggregates multiple sections from the INP / RPT into Dataframe and Geodataframe objects. With that, we can quickly visualize models with the familiar matplotlib API. For example:

```python 
# grab an example model 
from swmmio.examples import jersey

# do some matplotlib via geopandas
ax = jersey.links.geodataframe.plot(linewidth=jersey.links.dataframe['Geom1']*3, capstyle='round')
jersey.nodes.geodataframe.plot('MaxDepth', ax=ax, markersize='InvertElev', zorder=2)
jersey.subcatchments.geodataframe.plot('PercImperv', ax=ax, alpha=0.5, zorder=1)
```
![image](https://github.com/user-attachments/assets/545a7822-f668-4a0a-8166-b8b8321156e0)

Similarly, we can export model elements as GeoJSON from this API: 

```python 
jersey.nodes.geodataframe.to_json()
```
```
'{"type": "FeatureCollection", "features": [{"id": "J3", "type": "Feature", "properties": {"InvertElev": 6.547, "MaxDepth": 15.0, ...
```

Having said that, I still feel like it would be nice to make it easier to quickly visualize SWMM models, as you suggest. I like the idea of adding a `plot` method on the `swmmio.Model` object that returns a matplotlib object, and/or a folium-like interactive map that renders in Jupyter. My feeling is that we could approach this by leveraging the `swmmio.elements` module as described above (e.g. model.links.geodataframe.plot(), model.nodes.geodataframe.plot(), etc.)

What do you think of that? Overall, I definitely support you submitting a PR that helps move swmmio in this direction!

**Reply by everettsp** [reply]
Thanks for the detailed answer - some really good points. Everything said, I think a geopandas-based plotting and explore options would be useful features.

Currently I'm fetching some 'default' plotting args in a YAML file, e.g.:
```
nodes:
  junctions:
    zorder: 20
    color: blue
    linewidth: 2
  outfalls:
    zorder: 22
    color: red
    linewidth: 2
```
Simply iterate over that and plot with `plot()` or `explore()`. One annoying thing I ran into is that it'd be nice to have the type of each entry in swmmio.elements (i.e., whether a node is an outfall, junction, etc.), which allows a different style to be applied to each subtype of element. Currently, I'm just running this, but it's a bit inelegant:

```
from swmmio.defs import COMPOSITE_OBJECTS

def get_node_types(model:swmmio.Model) -> dict[str:str]:
    node_types = {}
    for sec in COMPOSITE_OBJECTS['nodes']["inp_sections"]:
        node_types.update({x:sec for x in getattr(model.inp, sec.lower()).index.tolist()})
    return node_types
```
Is there currently a nicer way of fetching this info, short of adding a new 'subtype' column to the swmmio.elements geodataframes?

Anyway, explore makes for a nice UI. If I'm feeling ambitious, I might tidy up some interactive timeseries results plotting tools, which I can eventually share as well.

![image](https://github.com/user-attachments/assets/7e03f2ad-6ea4-4a50-8dbf-1b2821076de8)
